# Proyecto 1 - Analítica de contratación pública

## Secretaría Distrital de Movilidad

**Integrante:** Manuela Torres

### Pregunta de negocio

¿Qué modalidades y tipos de contratación concentran la mayor parte del valor contratado por la Secretaría Distrital de Movilidad, y cómo ha cambiado esta composición a lo largo del tiempo?

### Objetivo

Realizar la exploración, limpieza y análisis de los datos de contratación de la Secretaría Distrital de Movilidad necesarios para responder la pregunta de negocio.

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("../Datos_completos_SDM.csv")

In [3]:
df.head()

,:id,:version,:created_at,:updated_at,nombre_entidad,nit_entidad,departamento,ciudad,localizaci_n,orden,...,n_mero_de_documento_ordenador_del_gasto,nombre_supervisor,tipo_de_documento_supervisor,n_mero_de_documento_supervisor,nombre_ordenador_de_pago,tipo_de_documento_ordenador_de_pago,n_mero_de_documento_ordenador_de_pago,documentos_tipo,descripcion_documentos_tipo,direcci_n_de_ejecuci_n_del_contrato
0,row-rd8p.mccw_35as,rv-4h3c_j643~kbx3,2026-08-23T08:01:59.248Z,2026-08-23T08:01:59.248Z,SECRETARÍA DISTRITAL DE MOVILIDAD,8999990619,Distrito Capital de Bogotá,Bogotá,"Colombia, Bogotá, Bogotá",Territorial,...,No definido,No definido,No definido,No definido,No definido,No definido,No definido,No,No definido,CALLE 13 # 37 - 35 Bogotá Distrito Capital de ...
1,row-ytp4.i46b.krcd,rv-tc43~4yf6-pdh7,2026-08-23T08:01:59.248Z,2026-08-23T08:01:59.248Z,SECRETARÍA DISTRITAL DE MOVILIDAD,8999990619,Distrito Capital de Bogotá,Bogotá,"Colombia, Bogotá, Bogotá",Territorial,...,No definido,No definido,No definido,No definido,No definido,No definido,No definido,No,No definido,Calle 13 #37 35 Bogotá Distrito Capital de Bog...
2,row-ukeb-eb8p-rxjx,rv-58ce~cnfa-jfgw,2026-08-23T08:01:59.248Z,2026-08-23T08:01:59.248Z,SECRETARÍA DISTRITAL DE MOVILIDAD,8999990619,Distrito Capital de Bogotá,Bogotá,"Colombia, Bogotá, Bogotá",Territorial,...,No definido,No definido,No definido,No definido,No definido,No definido,No definido,No,No definido,Calle 13 #37 35 Bogotá Distrito Capital de Bog...
3,row-776s.tfcx.qc3k,rv-bde8~rtvj~t5du,2026-08-23T08:01:59.248Z,2026-08-23T08:01:59.248Z,SECRETARÍA DISTRITAL DE MOVILIDAD,8999990619,Distrito Capital de Bogotá,Bogotá,"Colombia, Bogotá, Bogotá",Territorial,...,1026558244,No definido,No definido,No definido,No definido,No definido,No definido,No,No definido,Calle 13 #37 35 Bogotá Distrito Capital de Bog...
4,row-24jh-9jqg~66bc,rv-bsfk.c598.89ej,2026-08-23T08:01:59.248Z,2026-08-23T08:01:59.248Z,SECRETARÍA DISTRITAL DE MOVILIDAD,8999990619,Distrito Capital de Bogotá,Bogotá,"Colombia, Bogotá, Bogotá",Territorial,...,No definido,No definido,No definido,No definido,No definido,No definido,No definido,No,No definido,Cl 13 No 37-35 Bogotá Distrito Capital de Bogo...


In [4]:
df.shape

(21964, 89)

In [5]:
df.columns.tolist()

[':id',
 ':version',
 ':created_at',
 ':updated_at',
 'nombre_entidad',
 'nit_entidad',
 'departamento',
 'ciudad',
 'localizaci_n',
 'orden',
 'sector',
 'rama',
 'entidad_centralizada',
 'proceso_de_compra',
 'id_contrato',
 'referencia_del_contrato',
 'estado_contrato',
 'codigo_de_categoria_principal',
 'descripcion_del_proceso',
 'tipo_de_contrato',
 'modalidad_de_contratacion',
 'justificacion_modalidad_de',
 'fecha_de_firma',
 'fecha_de_inicio_del_contrato',
 'fecha_de_fin_del_contrato',
 'condiciones_de_entrega',
 'tipodocproveedor',
 'documento_proveedor',
 'proveedor_adjudicado',
 'es_grupo',
 'es_pyme',
 'habilita_pago_adelantado',
 'liquidaci_n',
 'obligaci_n_ambiental',
 'obligaciones_postconsumo',
 'reversion',
 'origen_de_los_recursos',
 'destino_gasto',
 'valor_del_contrato',
 'valor_de_pago_adelantado',
 'valor_facturado',
 'valor_pendiente_de_pago',
 'valor_pagado',
 'valor_amortizado',
 'valor_pendiente_de',
 'valor_pendiente_de_ejecucion',
 'saldo_cdp',
 'saldo_vige

## 1. Selección de variables relevantes

Para responder la pregunta de negocio se seleccionan las variables relacionadas con la identificación del contrato, su estado, el tipo y modalidad de contratación, la fecha de firma y el valor contractual.

In [6]:
columnas_interes = [
    "id_contrato",
    "estado_contrato",
    "tipo_de_contrato",
    "modalidad_de_contratacion",
    "fecha_de_firma",
    "valor_del_contrato"
]

df_manuela = df[columnas_interes].copy()

In [7]:
df_manuela.head()

,id_contrato,estado_contrato,tipo_de_contrato,modalidad_de_contratacion,fecha_de_firma,valor_del_contrato
0,CO1.PCCNTR.484426,Aprobado,Prestación de servicios,Contratación directa,2018-07-16T00:00:00.000,29274000.0
1,CO1.PCCNTR.7507462,Modificado,Prestación de servicios,Contratación directa,2025-02-17T00:00:00.000,58662000.0
2,CO1.PCCNTR.9585745,En ejecución,Prestación de servicios,Contratación directa,2026-06-22T00:00:00.000,58399400.0
3,CO1.PCCNTR.8779592,En ejecución,Prestación de servicios,Contratación directa,2026-01-07T00:00:00.000,29822100.0
4,CO1.PCCNTR.1700821,En ejecución,Prestación de servicios,Contratación directa,2020-07-16T00:00:00.000,20574250.0


In [8]:
df_manuela.shape

(21964, 6)

## 2. Revisión de calidad de los datos

Se revisan los tipos de datos, valores faltantes, categorías y posibles registros duplicados antes de realizar cualquier transformación o análisis.

In [9]:
df_manuela.info()

<class 'pandas.DataFrame'>
RangeIndex: 21964 entries, 0 to 21963
Data columns (total 6 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   id_contrato                21964 non-null  str    
 1   estado_contrato            21964 non-null  str    
 2   tipo_de_contrato           21964 non-null  str    
 3   modalidad_de_contratacion  21964 non-null  str    
 4   fecha_de_firma             21380 non-null  str    
 5   valor_del_contrato         21964 non-null  float64
dtypes: float64(1), str(5)
memory usage: 1.0 MB


In [10]:
df_manuela.isna().sum()

id_contrato                    0
estado_contrato                0
tipo_de_contrato               0
modalidad_de_contratacion      0
fecha_de_firma               584
valor_del_contrato             0
dtype: int64

In [11]:
(df_manuela.isna().mean() * 100).round(2)

id_contrato                  0.00
estado_contrato              0.00
tipo_de_contrato             0.00
modalidad_de_contratacion    0.00
fecha_de_firma               2.66
valor_del_contrato           0.00
dtype: float64

### Valores faltantes

La única variable seleccionada que presenta valores faltantes es `fecha_de_firma`, con 584 registros, equivalentes al 2,66 % de las observaciones.

Estos registros se conservarán para los análisis agregados que no requieren información temporal, como la distribución del valor contratado por tipo y modalidad. Para los análisis de evolución a lo largo del tiempo se utilizarán únicamente los contratos que cuentan con fecha de firma.

In [12]:
df_manuela["estado_contrato"].value_counts()

estado_contrato
Cerrado              7343
Modificado           5837
En ejecución         5600
Aprobado             1031
cedido                789
terminado             763
Borrador              384
Cancelado             150
enviado Proveedor      30
En aprobación          20
Suspendido             17
Name: count, dtype: int64

In [13]:
df_manuela["tipo_de_contrato"].value_counts()

tipo_de_contrato
Prestación de servicios       21348
Compraventa                     186
Otro                            143
Obra                             63
Interventoría                    50
Consultoría                      48
Suministros                      34
Seguros                          34
Decreto 092 de 2017              25
Arrendamiento de inmuebles       21
Comodato                          5
Arrendamiento de muebles          3
Servicios financieros             2
Concesión                         1
Comisión                          1
Name: count, dtype: int64

In [14]:
df_manuela["modalidad_de_contratacion"].value_counts()

modalidad_de_contratacion
Contratación directa                                           21110
Mínima cuantía                                                   257
Contratación Directa (con ofertas)                               140
Selección abreviada subasta inversa                              137
Concurso de méritos abierto                                       98
Licitación pública                                                85
Selección Abreviada de Menor Cuantía                              51
Licitación pública Obra Publica                                   50
Contratación régimen especial                                     28
Seleccion Abreviada Menor Cuantia Sin Manifestacion Interes        8
Name: count, dtype: int64

### Revisión de categorías

Se revisó la distribución de los estados, tipos y modalidades de contratación. Se observa una alta concentración en contratos de prestación de servicios y en la modalidad de contratación directa. Sin embargo, la frecuencia de contratos no necesariamente implica concentración del valor contratado, por lo que estas categorías se conservarán para su posterior análisis económico.

Antes de definir filtros sobre el estado contractual, se analizará la relación entre los estados y la disponibilidad de la fecha de firma.

In [15]:
df_manuela[df_manuela["fecha_de_firma"].isna()]["estado_contrato"].value_counts()

estado_contrato
Borrador             384
Cancelado            150
enviado Proveedor     30
En aprobación         20
Name: count, dtype: int64

In [16]:
df_manuela["id_contrato"].duplicated().sum()

np.int64(0)

In [17]:
df_manuela.duplicated().sum()

np.int64(0)

### Tratamiento de estados contractuales

Los 584 registros sin fecha de firma corresponden exclusivamente a los estados
`Borrador`, `Cancelado`, `enviado Proveedor` y `En aprobación`.

De acuerdo con el flujo de gestión contractual de SECOP II, los estados asociados al
envío al proveedor y a la aprobación corresponden a etapas previas a la firma y
publicación definitiva del contrato. Por esta razón, estos registros no se consideran
contratación formalizada y se excluyen del análisis del valor contratado.

También se verificó la existencia de duplicados tanto por `id_contrato` como por fila
completa, sin encontrarse registros duplicados.

In [18]:
estados_excluir = [
    "Borrador",
    "Cancelado",
    "enviado Proveedor",
    "En aprobación"
]

In [19]:
df_limpio = df_manuela[
    ~df_manuela["estado_contrato"].isin(estados_excluir)
].copy()

In [20]:
df_limpio.shape

(21380, 6)

In [21]:
df_limpio.isna().sum()

id_contrato                  0
estado_contrato              0
tipo_de_contrato             0
modalidad_de_contratacion    0
fecha_de_firma               0
valor_del_contrato           0
dtype: int64

### Revisión del valor del contrato

Antes de realizar el análisis económico se revisa la distribución de `valor_del_contrato` para identificar valores iguales a cero, negativos o potencialmente atípicos que puedan distorsionar los resultados agregados.

In [22]:
df_limpio["valor_del_contrato"].describe()

count    2.138000e+04
mean     1.510513e+08
std      1.358131e+09
min      0.000000e+00
25%      2.029000e+07
50%      2.982210e+07
75%      5.461000e+07
max      7.517333e+10
Name: valor_del_contrato, dtype: float64

In [23]:
(df_limpio["valor_del_contrato"] <= 0).sum()

np.int64(24)

In [24]:
df_limpio[
    ["id_contrato", "tipo_de_contrato", "modalidad_de_contratacion", "valor_del_contrato"]
].sort_values(
    "valor_del_contrato",
    ascending=False
).head(10)

,id_contrato,tipo_de_contrato,modalidad_de_contratacion,valor_del_contrato
11395,CO1.PCCNTR.8559979,Otro,Contratación Directa (con ofertas),7.517333e+10
14280,CO1.PCCNTR.7918128,Otro,Contratación Directa (con ofertas),4.615501e+10
2395,CO1.PCCNTR.2412371,Prestación de servicios,Licitación pública,4.488798e+10
5910,CO1.PCCNTR.7973611,Prestación de servicios,Licitación pública,3.845067e+10
21328,CO1.PCCNTR.6387308,Otro,Contratación Directa (con ofertas),3.519287e+10
4389,CO1.PCCNTR.7770910,Prestación de servicios,Selección abreviada subasta inversa,3.302960e+10
3370,CO1.PCCNTR.2432617,Decreto 092 de 2017,Contratación régimen especial,3.185929e+10
10391,CO1.PCCNTR.4679275,Obra,Licitación pública Obra Publica,2.832564e+10
19213,CO1.PCCNTR.7766240,Obra,Licitación pública Obra Publica,2.777025e+10
15163,CO1.PCCNTR.8716810,Prestación de servicios,Licitación pública,2.734032e+10


In [25]:
(df_limpio["valor_del_contrato"] == 0).sum()

np.int64(24)

In [26]:
(df_limpio["valor_del_contrato"] < 0).sum()

np.int64(0)

In [27]:
df_limpio[
    df_limpio["valor_del_contrato"] <= 0
][
    ["id_contrato", "estado_contrato", "tipo_de_contrato",
     "modalidad_de_contratacion", "fecha_de_firma",
     "valor_del_contrato"]
]

,id_contrato,estado_contrato,tipo_de_contrato,modalidad_de_contratacion,fecha_de_firma,valor_del_contrato
2169,CO1.PCCNTR.1490924,terminado,Consultoría,Concurso de méritos abierto,2020-04-13T00:00:00.000,0.0
4555,CO1.PCCNTR.990319,terminado,Consultoría,Concurso de méritos abierto,2019-06-10T00:00:00.000,0.0
4608,CO1.PCCNTR.1849461,terminado,Otro,Contratación Directa (con ofertas),2020-09-18T00:00:00.000,0.0
4693,CO1.PCCNTR.5146541,Modificado,Otro,Contratación directa,2023-06-27T00:00:00.000,0.0
6033,CO1.PCCNTR.3020725,Modificado,Otro,Contratación directa,2021-12-06T00:00:00.000,0.0
6192,CO1.PCCNTR.2755044,Modificado,Otro,Contratación directa,2021-08-12T00:00:00.000,0.0
6555,CO1.PCCNTR.3833493,terminado,Otro,Contratación directa,2022-07-22T00:00:00.000,0.0
6718,CO1.PCCNTR.2733162,Aprobado,Otro,Contratación directa,2021-08-05T00:00:00.000,0.0
7494,CO1.PCCNTR.2348603,terminado,Comodato,Contratación directa,2021-03-15T00:00:00.000,0.0
8596,CO1.PCCNTR.1523960,Modificado,Otro,Contratación directa,2020-04-24T00:00:00.000,0.0


### Tratamiento de contratos con valor cero

Se identificaron 24 contratos con `valor_del_contrato` igual a cero y no se encontraron valores negativos.

Los contratos con valor cero corresponden a distintos tipos, modalidades y estados contractuales, por lo que con la información seleccionada no es posible concluir que se trate de registros erróneos. Por esta razón, se conservan dentro de la base limpia.

Estos registros no afectan los análisis basados en la suma del valor contratado. Sin embargo, para análisis de distribución del monto contractual, como promedios, diagramas de caja o pruebas estadísticas, se trabajará únicamente con contratos cuyo valor sea mayor que cero.

In [28]:
df_monetario = df_limpio[
    df_limpio["valor_del_contrato"] > 0
].copy()

In [29]:
df_monetario.shape

(21356, 6)

### Transformación de la variable temporal

La variable `fecha_de_firma` se transforma a formato de fecha para permitir el análisis temporal. A partir de esta variable se crea `año`, que permitirá estudiar la evolución anual de la contratación.

In [30]:
df_limpio["fecha_de_firma"] = pd.to_datetime(
    df_limpio["fecha_de_firma"]
)

df_limpio["año"] = df_limpio["fecha_de_firma"].dt.year

In [31]:
df_monetario["fecha_de_firma"] = pd.to_datetime(
    df_monetario["fecha_de_firma"]
)

df_monetario["año"] = df_monetario["fecha_de_firma"].dt.year

In [32]:
df_limpio.info()

<class 'pandas.DataFrame'>
Index: 21380 entries, 0 to 21963
Data columns (total 7 columns):
 #   Column                     Non-Null Count  Dtype         
---  ------                     --------------  -----         
 0   id_contrato                21380 non-null  str           
 1   estado_contrato            21380 non-null  str           
 2   tipo_de_contrato           21380 non-null  str           
 3   modalidad_de_contratacion  21380 non-null  str           
 4   fecha_de_firma             21380 non-null  datetime64[us]
 5   valor_del_contrato         21380 non-null  float64       
 6   año                        21380 non-null  int32         
dtypes: datetime64[us](1), float64(1), int32(1), str(4)
memory usage: 1.2 MB


In [33]:
df_limpio["año"].value_counts().sort_index()

año
2017      19
2018    1758
2019     461
2020    2009
2021    2909
2022    2056
2023    2729
2024    3583
2025    3280
2026    2576
Name: count, dtype: int64